sql_script/MS3_aggregation.sql

```diff
@@ -1,34 +1,36 @@
-/*(NESTED AGGREGATION) for each warehouse, find the youngest employee who works there -- NOT TO CONFIDENT ON THIS */
-SELECT W.id, P.id, P.first_name, P.last_name
+/*(NESTED AGGREGATION) for each warehouse, find the youngest employee who works there 
+POSTGRESQL-SPECIFIC ISSUE!*/
```

Comment:
NEED HELP DEBUGGING:
PostgreSQL-specific problem: 
- "P.first_name must appear in the GROUP BY clause or be used in an aggregate function"
- "P.last_name must appear in the GROUP BY clause or be used in an aggregate function"
=> All attributes in GROUP BY must also be in SELECT, but that doesn't work for what we need to do.

Reply:
I don't think this is a postgres problem. The slides actually say the inverse, that all attributes in the SELECT must also be in the GROUP BY.

![image](https://user-images.githubusercontent.com/3186109/99290966-4f438900-27fc-11eb-9f2e-5045153b46be.png)

Reply:
I'm having a hard time figuring out how to do this though... If you just want the minimum age in each warehouse, you'd do:

```sql
SELECT W.id, MIN(P.birth_date)
FROM warehouse W, personnel P, works_in N
WHERE W.id = N.warehouse_id AND
      P.id = N.pid
GROUP BY W.id
HAVING COUNT(*) >= 1;
```

But that doesn't give you the specific student. I'm not sure if that's even possible actually, since the aggregation operators just return the result of doing something on a big batch of data (as per their name).

If it were possible though, I think the check would need to happen in `HAVING`. Right now you're checking in `WHERE`, but that filtering gets applied before the `GROUP BY` happens. i.e., you're pre-filtering down the data to only the oldest employees (that should probably be `MIN`).

Reply:
You could do this, but I don't think this fits the criteria of nested aggregation:

```sql
SELECT N.warehouse_id, P.id, P.first_name, P.last_name, P.birth_date
FROM personnel P, works_in N
WHERE P.id = N.pid AND
     (N.warehouse_id, P.birth_date) in
    (SELECT W.id, MIN(P.birth_date)
    FROM warehouse W, personnel P, works_in N
    WHERE W.id = N.warehouse_id AND
          P.id = N.pid
    GROUP BY W.id
    HAVING COUNT(*) >= 1);
```

Reply:
Yah there's definitely a way to do it without nesting, but then it doesn't meet the criterion. We might need to come up with a different query entirely.

Reply:
What about the inventory types with larger than average volume?

```sql
SELECT type_name
FROM inventory_type IT
GROUP BY type_name
HAVING (height * width * length) > (
  SELECT AVG(height * width * length)
  FROM inventory_type);
```

This only produces `TY Bat Mobile Stuffy` though since that item pulls up the average so high, and likewise the opposite produces all items except for `TY Bat Mobile Stuffy`. :)

Reply:
What about the inventory types with smaller than average volume + the number of those items?

```sql
SELECT IT.type_name, COUNT(*)
FROM inventory_type IT, inventory I
WHERE I.type_name = IT.type_name
GROUP BY IT.type_name
HAVING (IT.height * IT.width * IT.length) < (
  SELECT AVG(IT.height * IT.width * IT.length)
  FROM inventory_type IT);
```

```txt
             type_name             | count
-----------------------------------+-------
 Fidget Spinner                    |     1
 Intel Core i7-6700 0.02 Processor |     1
 B/W striped notebook              |     2
 HB Graphite Pencil                |     1
(4 rows)
```